### PR TITLE
fix upgradeable contracts and add unit tests

### PR DESCRIPTION
**Pull request by rzrrbk**
### Description
According to the [contract desigin](https://github.com/bnb-chain/zkbnb-contract/blob/master/CONTRACTS_DESIGN.md#upgradeable-design) there are 5 contracts which are upgradeable. 

This PR fixes contracts bug and adds unit tests to make sure the upgrade works.

### Example

To run the upgradeable-releated unit tests, simply execute:
```shell
$ yarn test-upgradeable
```
The output be like:
```
yarn test-upgradeable
yarn run v1.22.19
warning ../../package.json: No license field
$ npx hardhat test test/Proxy.test.js test/UpgradeableMaster.test.js  test/UpgradeGatekeeper.test.js


  Proxy
    ✔ Proxy contract should initialize target contract
    ✔ Proxy contract should store target address 
    Proxy contract should upgrade new target
      ✔ upgrade new `Governance` target (104ms)
      ✔ upgrade new `ZkBNBVerifier` target (127ms)
      ✔ upgrade new `ZNSController` target (108ms)
      ✔ upgrade new `PublicResolver` target (110ms)
      ✔ upgrade new `ZkBNB` target (113ms)
    Proxy contract should delegate function
      ✔ delegate `changeGovernor` function
      ✔ delegate `verifyProof` function
      ✔ delegate `registerZNS` function
      ✔ delegate `name` function
      ✔ delegate `commitBlocks` function
    Proxy contract should intercept upgrade function
      ✔ intercept `Governance` upgrade
      ✔ intercept `ZkBNBVerifier` upgrade
      ✔ intercept `ZNSController` upgrade
      ✔ intercept `PublicResolver` upgrade
      ✔ intercept `ZkBNB` upgrade

  UpgradeableMaster
    Notice period
      ✔ Default notice period should be the shortest
      ✔ Preparation status should be activated 4 weeks after notification started
      ✔ Notice period can be cut by council members (43ms)
      ✔ Ready for upgrade
      Clear upgrade status
        ✔ Clear upgrade status when cancelled
        ✔ Clear upgrade status when finishes
      Access control
        - Only council member can invoke `upgradeNoticePeriodStarted`

  UpgradeGatekeeper
    Add new upgradeable contracts
      ✔ added `proxyGovernance`
      ✔ added `proxyZkBNBVerifier`
      ✔ added `proxyZNSController`
      ✔ added `proxyPublicResolver`
      ✔ added `proxyZkBNB`
    Check normal upgrade
      ✔ start upgrade
      ✔ start preparation
      ✔ finish upgrade
    Check upgrade cancellation
      ✔ should fail to cancel before upgrade
      ✔ cancel after start upgrade
      ✔ cancel after start preparation
      ✔ should fail to cancel after finish upgrade (41ms)


  35 passing (14s)
  1 pending

✨  Done in 16.33s.
```
### Changes

Notable changes:
* refactored `DeployFactory` to avoid stack too deep error
* added missing `upgrade` function `ZNSController`, `PublicResolver` and `ZkBNB` contract
* added unit tests for `Proxy`, `UpgradeableMaster` and `UpgradeGatekeeper`
* fixed lint error

### CODE REVIEW

File: CONTRACTS_DESIGN.md [hunk 1]
```diff
@@ -234,6 +234,8 @@ as the resolver for his nodes. Then others can resolve this name for detailed in
 
 
 ## Upgradeable Design
+
+<!-- TODO: upgrade this doc -->
```
Comment (galler-dev) [inline CONTRACTS_DESIGN.md:238]: Need update doc

Comment (JonneyTheSheep) [inline CONTRACTS_DESIGN.md]: Documentation needs to be updated

File: package.json [hunk 2]
```diff
@@ -7,13 +7,15 @@
     "build": "hardhat compile",
     "prebuild": "rimraf ./artifacts/ ./cache/",
     "test": "echo \"Error: no test specified\" && exit 1",
+    "test-upgradeable": "npx hardhat test test/Proxy.test.js test/UpgradeableMaster.test.js  test/UpgradeGatekeeper.test.js",
```
Comment (JonneyTheSheep) [inline package.json]:
```suggestion
    "test-upgradeable": "npx hardhat test test/upgradeable/**.test.js
```

File: CONTRACTS_DESIGN.md (hunk 1 above)
Comment (rzrrbk) [reply CONTRACTS_DESIGN.md]: updated 🚀

File: package.json (hunk 2 above)
Comment (rzrrbk) [reply package.json]: done 🦜

**Review by JonneyTheSheep — COMMENTED** [review]
/lgtm

**Review by qingyang-id — APPROVED** [review]

**Pull request merged**